ignite/internal/tools/gen-mig-diffs/pkg/repo/repo.go

```diff
@@ -0,0 +1,384 @@
+)
+
+const (
+	defaultRepoURL    = "https://github.com/ignite/cli.git"
+	defaultBinaryPath = "dist/ignite"
+)
+
+type (
+	// Generator is used to generate migration diffs.
+	Generator struct {
+		From, To *semver.Version
+		source   string
+		binPath  string
+		repo     *git.Repository
+		session  *cliui.Session
+		cleanup  bool
+	}
+
+	// options represents configuration for the generator.
+	options struct {
+		source  string
+		output  string
+		repoURL string
+		binPath string
+		cleanup bool
+	}
+	// Options configures the generator.
+	Options func(*options)
+)
+
+// newOptions returns a options with default options.
+func newOptions() options {
+	tmpDir := os.TempDir()
+	return options{
+		source:  "",
+		binPath: filepath.Join(tmpDir, "bin"),
+		output:  filepath.Join(tmpDir, "migration-source"),
+		repoURL: defaultRepoURL,
+	}
+}
+
+// WithSource set the repo source Options.
+func WithSource(source string) Options {
+	return func(o *options) {
+		o.source = source
+		// Do not clean up if set the source.
+		o.cleanup = false
+	}
+}
+
+// WithRepoURL set the repo URL Options.
+func WithRepoURL(repoURL string) Options {
+	return func(o *options) {
+		o.repoURL = repoURL
+	}
+}
+
+// WithRepoOutput set the repo output Options.
+func WithRepoOutput(output string) Options {
+	return func(o *options) {
+		o.output = output
+	}
+}
+
+// WithBinPath set the binary path to build the source.
+func WithBinPath(binPath string) Options {
+	return func(o *options) {
+		o.binPath = binPath
+	}
+}
+
+// WithCleanup cleanup folders after use.
+func WithCleanup() Options {
+	return func(o *options) {
+		o.cleanup = true
+	}
+}
+
+// validate options.
+func (o options) validate() error {
+	if o.source != "" && (o.repoURL != defaultRepoURL) {
+		return errors.New("cannot set source and repo URL at the same time")
+	}
+	if o.source != "" && o.cleanup {
+		return errors.New("cannot set source and cleanup the repo")
+	}
+	return nil
+}
+
+// New creates a new generator for migration diffs between from and to versions of ignite cli
+// If source is empty, then it clones the ignite cli repository to a temporary directory and uses it as the source.
+func New(from, to *semver.Version, session *cliui.Session, options ...Options) (*Generator, error) {
+	opts := newOptions()
+	for _, apply := range options {
+		apply(&opts)
+	}
+
+	err := opts.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		source = opts.source
+		repo   *git.Repository
+	)
+	if source != "" {
+		repo, err = git.PlainOpen(source)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to open ignite repository")
+		}
+
+		session.StopSpinner()
+		session.EventBus().SendInfo(fmt.Sprintf("Using ignite repository at: %s", source))
+	} else {
+		session.StartSpinner("Cloning ignite repository...")
+
+		source = opts.output
+		if err := os.RemoveAll(source); err != nil {
+			return nil, errors.Wrap(err, "failed to clean the output directory")
+		}
```

Comment:
Maybe the output directory should be clean already? Otherwise this could unintentionally remove the wrong folder if the given path is wrong. It has low provability of happening given that we will use it internally but I thought it was worth mentioning, just in case.